File: pkg/feature/domain/clause_evaluator.go
// Copyright 2024 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package domain

import (
	"strconv"
	"strings"

	"github.com/blang/semver"

	featureproto "github.com/bucketeer-io/bucketeer/proto/feature"
)

type clauseEvaluator struct {
	segmentEvaluator
}

func (c *clauseEvaluator) Evaluate(
	targetValue string,
	clause *featureproto.Clause,
	userID string,
	segmentUsers []*featureproto.SegmentUser,
) bool {
	switch clause.Operator {
	case featureproto.Clause_EQUALS:
		// TODO: this should only be one value or equals makes no sense.
		return c.equals(targetValue, clause.Values)
	case featureproto.Clause_IN:
		return c.in(targetValue, clause.Values)
	case featureproto.Clause_STARTS_WITH:
		return c.startsWith(targetValue, clause.Values)
	case featureproto.Clause_ENDS_WITH:
		return c.endsWith(targetValue, clause.Values)
	case featureproto.Clause_SEGMENT:
		return c.segmentEvaluator.Evaluate(clause.Values, userID, segmentUsers)
	case featureproto.Clause_GREATER:
		return c.greater(targetValue, clause.Values)
	case featureproto.Clause_GREATER_OR_EQUAL:
		return c.greaterOrEqual(targetValue, clause.Values)
	case featureproto.Clause_LESS:
		return c.less(targetValue, clause.Values)
	case featureproto.Clause_LESS_OR_EQUAL:
		return c.lessOrEqual(targetValue, clause.Values)
	case featureproto.Clause_BEFORE:
		return c.before(targetValue, clause.Values)
	case featureproto.Clause_AFTER:
		return c.after(targetValue, clause.Values)
	}
	return false
}

func (c *clauseEvaluator) equals(targetValue string, values []string) bool {
	for i := range values {
		if targetValue == values[i] {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) in(targetValue string, values []string) bool {
	for i := range values {
		if targetValue == values[i] {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) startsWith(targetValue string, values []string) bool {
	for i := range values {
		if strings.HasPrefix(targetValue, values[i]) {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) endsWith(targetValue string, values []string) bool {
	for i := range values {
		if strings.HasSuffix(targetValue, values[i]) {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) greater(targetValue string, values []string) bool {
	floatTarget, floatValues, err := parseFloat(targetValue, values)
	if err == nil {
		for _, value := range floatValues {
			if floatTarget > value {
				return true
			}
		}
		return false
	}
	semverTarget, semverValues, err := parseSemver(targetValue, values)
	if err == nil {
		for _, value := range semverValues {
			if semverTarget.GT(value) {
				return true
			}
		}
		return false
	}
	for _, value := range values {
		if targetValue > value {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) greaterOrEqual(targetValue string, values []string) bool {
	floatTarget, floatValues, err := parseFloat(targetValue, values)
	if err == nil {
		for _, value := range floatValues {
			if floatTarget >= value {
				return true
			}
		}
		return false
	}
	semverTarget, semverValues, err := parseSemver(targetValue, values)
	if err == nil {
		for _, value := range semverValues {
			if semverTarget.GTE(value) {
				return true
			}
		}
		return false
	}
	for _, value := range values {
		if targetValue >= value {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) less(targetValue string, values []string) bool {
	floatTarget, floatValues, err := parseFloat(targetValue, values)
	if err == nil {
		for _, value := range floatValues {
			if floatTarget < value {
				return true
			}
		}
		return false
	}
	semverTarget, semverValues, err := parseSemver(targetValue, values)
	if err == nil {
		for _, value := range semverValues {
			if semverTarget.LT(value) {
				return true
			}
		}
		return false
	}
	for _, value := range values {
		if targetValue < value {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) lessOrEqual(targetValue string, values []string) bool {
	floatTarget, floatValues, err := parseFloat(targetValue, values)
	if err == nil {
		for _, value := range floatValues {
			if floatTarget <= value {
				return true
			}
		}
		return false
	}
	semverTarget, semverValues, err := parseSemver(targetValue, values)
	if err == nil {
		for _, value := range semverValues {
			if semverTarget.LTE(value) {
				return true
			}
		}
		return false
	}
	for _, value := range values {
		if targetValue <= value {
			return true
		}
	}
	return false
}

func (c *clauseEvaluator) before(targetValue string, values []string) bool {
	intTarget, intValues, err := parseInt(targetValue, values)
	if err == nil {
		for _, value := range intValues {
			if intTarget < value {
				return true
			}
		}
	}
	return false
}

func (c *clauseEvaluator) after(targetValue string, values []string) bool {
	intTarget, intValues, err := parseInt(targetValue, values)
	if err == nil {
		for _, value := range intValues {
			if intTarget > value {
				return true
			}
		}
	}
	return false
}

func parseInt(targetValue string, values []string) (int64, []int64, error) {
	intTarget, err := strconv.ParseInt(targetValue, 10, 64)
	if err != nil {
		return -1, nil, err
	}
	intValues := make([]int64, 0, len(values))
	for _, value := range values {
		v, err := strconv.ParseInt(value, 10, 64)
		if err == nil {
			intValues = append(intValues, v)
		}
	}
	return intTarget, intValues, nil
}

func parseFloat(targetValue string, values []string) (float64, []float64, error) {
	floatTarget, err := strconv.ParseFloat(targetValue, 64)
	if err != nil {
		return -1, nil, err
	}
	floatValues := make([]float64, 0, len(values))
	for _, value := range values {
		v, err := strconv.ParseFloat(value, 64)
		if err == nil {
			floatValues = append(floatValues, v)
		}

	}
	return floatTarget, floatValues, nil
}

func parseSemver(targetValue string, values []string) (semver.Version, []semver.Version, error) {
	versionTarget, err := semver.Parse(targetValue)
	if err != nil {
		return semver.Version{}, nil, err
	}
	versionValues := make([]semver.Version, 0, len(values))
	for _, value := range values {
		v, err := semver.Parse(value)
		if err == nil {
			versionValues = append(versionValues, v)
		}
	}
	if err != nil {
		return semver.Version{}, nil, err
	}
	return versionTarget, versionValues, nil
}

File: evaluation/typescript/src/clauseEvaluator.ts
import { Clause } from './proto/feature/clause_pb';
import { SegmentUser } from './proto/feature/segment_pb';
import { SegmentEvaluator } from './segmentEvaluator';
import { DependencyEvaluator } from './dependencyEvaluator';
import * as semver from 'semver';
//
class ClauseEvaluator {
  private segmentEvaluator: SegmentEvaluator;
  private dependencyEvaluator: DependencyEvaluator;

  constructor() {
    this.segmentEvaluator = new SegmentEvaluator();
    this.dependencyEvaluator = new DependencyEvaluator();
  }

  evaluate(
    targetValue: string,
    clause: Clause,
    userID: string,
    segmentUsers: SegmentUser[],
    flagVariations: { [key: string]: string },
  ): boolean {
    try {
      switch (clause.getOperator()) {
        case Clause.Operator.EQUALS:
          return this.equals(targetValue, clause.getValuesList());
        case Clause.Operator.IN:
          return this.in(targetValue, clause.getValuesList());
        case Clause.Operator.STARTS_WITH:
          return this.startsWith(targetValue, clause.getValuesList());
        case Clause.Operator.ENDS_WITH:
          return this.endsWith(targetValue, clause.getValuesList());
        case Clause.Operator.SEGMENT:
          return this.segmentEvaluator.evaluate(clause.getValuesList(), userID, segmentUsers);
        case Clause.Operator.GREATER:
          return this.greater(targetValue, clause.getValuesList());
        case Clause.Operator.GREATER_OR_EQUAL:
          return this.greaterOrEqual(targetValue, clause.getValuesList());
        case Clause.Operator.LESS:
          return this.less(targetValue, clause.getValuesList());
        case Clause.Operator.LESS_OR_EQUAL:
          return this.lessOrEqual(targetValue, clause.getValuesList());
        case Clause.Operator.BEFORE:
          return this.before(targetValue, clause.getValuesList());
        case Clause.Operator.AFTER:
          return this.after(targetValue, clause.getValuesList());
        case Clause.Operator.FEATURE_FLAG:
          return this.dependencyEvaluator.evaluate(
            clause.getAttribute(),
            clause.getValuesList(),
            flagVariations,
          );
        case Clause.Operator.PARTIALLY_MATCH:
          return this.partiallyMatches(targetValue, clause.getValuesList());
        case Clause.Operator.NOT_EQUALS:
          return !this.equals(targetValue, clause.getValuesList());
        default:
          return false;
      }
    } catch (error) {
      console.error('Error evaluating clause:', error);
      throw error;
    }
  }

  private equals(targetValue: string, values: string[]): boolean {
    // First try semver comparison (with v-prefix normalization via semver.valid)
    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      const semverValues = values
        .map((v) => semver.valid(v))
        .filter((v): v is string => v !== null);
      // Target parsed as semver; if no values parsed or none matched, return false
      // (consistent with greaterOrEqual, greater, less, lessOrEqual behavior)
      return semverValues.some((value) => semver.eq(semverTarget, value));
    }
    // Fall back to exact string comparison only when target fails to parse as semver
    return values.includes(targetValue);
  }

  private partiallyMatches(targetValue: string, values: string[]): boolean {
    return values.some((value) => targetValue.includes(value));
  }

  private in(targetValue: string, values: string[]): boolean {
    // First try semver comparison (with v-prefix normalization via semver.valid)
    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      const semverValues = values
        .map((v) => semver.valid(v))
        .filter((v): v is string => v !== null);
      // Target parsed as semver; if no values parsed or none matched, return false
      // (consistent with greaterOrEqual, greater, less, lessOrEqual behavior)
      return semverValues.some((value) => semver.eq(semverTarget, value));
    }
    // Fall back to exact string comparison only when target fails to parse as semver
    return values.includes(targetValue);
  }

  private startsWith(targetValue: string, values: string[]): boolean {
    return values.some((value) => targetValue.startsWith(value));
  }

  private endsWith(targetValue: string, values: string[]): boolean {
    return values.some((value) => targetValue.endsWith(value));
  }

  private greater(targetValue: string, values: string[]): boolean {
    const floatTarget = validateAndParseFloat(targetValue);
    const floatValues = values.map(validateAndParseFloat).filter((value) => !isNaN(value));

    if (!isNaN(floatTarget)) {
      if (floatValues.length == 0) {
        return false;
      }
      return floatValues.some((value) => floatTarget > value);
    }

    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      return values.some((value) => {
        if (semver.valid(value)) {
          return semver.gt(semverTarget, value);
        }
        return false;
      });
    }

    return values.some((value) => targetValue > value);
  }

  private greaterOrEqual(targetValue: string, values: string[]): boolean {
    const floatTarget = validateAndParseFloat(targetValue);
    const floatValues = values.map(validateAndParseFloat).filter((value) => !isNaN(value));

    if (!isNaN(floatTarget)) {
      if (floatValues.length == 0) {
        return false;
      }
      return floatValues.some((value) => floatTarget >= value);
    }

    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      return values.some((value) => {
        if (semver.valid(value)) {
          return semver.gte(semverTarget, value);
        }
        return false;
      });
    }

    return values.some((value) => targetValue >= value);
  }

  private less(targetValue: string, values: string[]): boolean {
    const floatTarget = validateAndParseFloat(targetValue);
    const floatValues = values.map(validateAndParseFloat).filter((value) => !isNaN(value));

    if (!isNaN(floatTarget)) {
      if (floatValues.length == 0) {
        return false;
      }
      return floatValues.some((value) => floatTarget < value);
    }

    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      return values.some((value) => {
        if (semver.valid(value)) {
          return semver.lt(semverTarget, value);
        }
        return false;
      });
    }

    return values.some((value) => targetValue < value);
  }

  private lessOrEqual(targetValue: string, values: string[]): boolean {
    const floatTarget = validateAndParseFloat(targetValue);
    const floatValues = values.map(validateAndParseFloat).filter((value) => !isNaN(value));

    if (!isNaN(floatTarget)) {
      if (floatValues.length == 0) {
        return false;
      }
      return floatValues.some((value) => floatTarget <= value);
    }

    const semverTarget = semver.valid(targetValue);
    if (semverTarget) {
      return values.some((value) => {
        if (semver.valid(value)) {
          return semver.lte(semverTarget, value);
        }
        return false;
      });
    }

    return values.some((value) => targetValue <= value);
  }

  private before(targetValue: string, values: string[]): boolean {
    const intTarget = validateAndParseInt(targetValue);
    if (isNaN(intTarget)) {
      return false;
    }
    const intValues = values.map((value) => parseInt(value, 10)).filter((value) => !isNaN(value));
    if (intValues.length == 0) {
      return false;
    }

    return intValues.some((value) => intTarget < value);
  }

  private after(targetValue: string, values: string[]): boolean {
    const intTarget = parseInt(targetValue, 10);
    const intValues = values.map((value) => parseInt(value, 10)).filter((value) => !isNaN(value));

    return intValues.some((value) => intTarget > value);
  }
}

function isNumericString(input: string): boolean {
  return /^-?\d*\.?\d+$/.test(input);
}

function validateAndParseInt(input: string): number {
  // Check if the string is a valid number using regex
  if (!isNumericString(input)) {
    // Will not allow a string like `1a`, `0b`
    return NaN;
  }

  // Parse the string to float
  const result = parseInt(input, 10);

  // If parsing fails, return NaN
  return isNaN(result) ? NaN : result;
}

function validateAndParseFloat(input: string): number {
  // Check if the string is a valid number using regex
  if (!isNumericString(input)) {
    // Will not allow a string like `1a`, `0b`
    return NaN;
  }

  // Parse the string to float
  const result = parseFloat(input);

  // If parsing fails, return NaN
  return isNaN(result) ? NaN : result;
}

export { ClauseEvaluator };
